File: src/engine/EnigWindow.java
package engine;

import Game.UserControls;
import org.lwjgl.glfw.GLFWCursorPosCallback;
import org.lwjgl.glfw.GLFWErrorCallback;
import org.lwjgl.glfw.GLFWVidMode;
import org.lwjgl.opengl.GL;
import org.lwjgl.opengl.GL11;
import org.lwjgl.system.MemoryStack;

import java.nio.IntBuffer;

import static org.lwjgl.glfw.Callbacks.glfwFreeCallbacks;
import static org.lwjgl.glfw.GLFW.*;
import static org.lwjgl.opengl.ARBImaging.GL_TABLE_TOO_LARGE;
import static org.lwjgl.opengl.GL11.*;
import static org.lwjgl.opengl.GL11.GL_OUT_OF_MEMORY;
import static org.lwjgl.opengl.GL13.GL_MULTISAMPLE;
import static org.lwjgl.opengl.GL30.GL_INVALID_FRAMEBUFFER_OPERATION;
import static org.lwjgl.system.MemoryStack.stackPush;
import static org.lwjgl.system.MemoryUtil.NULL;

public class EnigWindow {
	// The window handle
	public long window;

	public static int width = 1800/1;
	public static int height = (int)(1100/1);

	public float cursorXFloat;
	public float cursorYFloat;

	public double cursorXDouble;
	public double cursorYDouble;

	public double cursorXOffset;
	public double cursorYOffset;

	public boolean inputEnabled = false;
	public boolean printFrames = false;

	public int[] keys = new int[GLFW_KEY_LAST + 1];
	
	public int framesSinceLastSecond = 0;
	public int lastFrameCount = 0;
	public long lastSecond;
	//public long lastFrame = System.nanoTime();
	
	public static EnigWindow mainWindow;

	public void terminate() {
		//System.out.println("Hello LWJGL " + Version.getVersion() + "!");

		//init();
		//loop();

		// Free the window callbacks and destroy the window
		glfwFreeCallbacks(window);
		glfwDestroyWindow(window);

		// Terminate GLFW and free the error callback
		glfwTerminate();
		glfwSetErrorCallback(null).free();
	}

	public EnigWindow() {
		// Setup an error callback. The default implementation
		// will print the error message in System.err.
		GLFWErrorCallback.createPrint(System.err).set();

		// Initialize GLFW. Most GLFW functions will not work before doing this.
		if ( !glfwInit() ) {
			throw new IllegalStateException("Unable to initialize GLFW");
		}

		// Configure GLFW
		glfwDefaultWindowHints(); // optional, the current window hints are already the default
		glfwWindowHint(GLFW_SAMPLES, 8);
		glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE); // the window will stay hidden after creation
		glfwWindowHint(GLFW_RESIZABLE, GLFW_TRUE); // the window will be resizable

		// Create the window
		GLFWVidMode vidMode = glfwGetVideoMode(glfwGetPrimaryMonitor());
		width = vidMode.width();
		height = vidMode.height();
		window = glfwCreateWindow(width, height, "Enignets game", NULL/*glfwGetPrimaryMonitor()*/, NULL);
		if ( window == NULL ) {
			throw new RuntimeException("Failed to create the GLFW window");
		}

		// Setup a key callback. It will be called every time a key is pressed, repeated or released.
		glfwSetKeyCallback(window, (window, key, scancode, action, mods) -> {
			if ( key == GLFW_KEY_ESCAPE && action == GLFW_RELEASE ) {
				glfwSetWindowShouldClose(window, true); // We will detect this in the rendering loop
			}else if (key == UserControls.pause && action == GLFW_RELEASE) {
				toggleCursorInput();
			}else {
				keys[key] = action;
			}
		});

		glfwSetCursorPosCallback(window, new GLFWCursorPosCallback() {
			@Override
			public void invoke(long window, double xpos, double ypos) {
				if (inputEnabled) {
					cursorXOffset += cursorXDouble - xpos;
					cursorYOffset += cursorYDouble - ypos;
					cursorXDouble = xpos;
					cursorYDouble = ypos;
					cursorXFloat = (float) (xpos - width / 2) / (float) width;
					cursorYFloat = (float) (ypos - height / 2) / (float) height;
				}
			}
		});


		// Get the thread stack and push a new frame
		try ( MemoryStack stack = stackPush() ) {
			IntBuffer pWidth = stack.mallocInt(1); // int*
			IntBuffer pHeight = stack.mallocInt(1); // int*

			// Get the window size passed to glfwCreateWindow
			glfwGetWindowSize(window, pWidth, pHeight);

			// Get the resolution of the primary monitor
			GLFWVidMode vidmode = glfwGetVideoMode(glfwGetPrimaryMonitor());

			// Center the window
			glfwSetWindowPos(
					window,
					(vidmode.width() - pWidth.get(0)) / 2,
					(vidmode.height() - pHeight.get(0)) / 2
			);
		} // the stack frame is popped automatically

		// Make the OpenGL context current
		glfwMakeContextCurrent(window);
		// Enable v-sync
		glfwSwapInterval(1);

		glfwSetInputMode(window, GLFW_CURSOR, GLFW_CURSOR_NORMAL);

		// Make the window visible
		glfwShowWindow(window);

		glfwSetWindowSizeCallback(window, (long window, int w, int h) -> {
			width = w;
			height = h;
			glViewport(0, 0, width, height);
		});

		GL.createCapabilities();
		glEnable(GL_BLEND);
		glEnable(GL_DEPTH_TEST);
		glEnable(GL_MULTISAMPLE);
		glEnable(GL_CULL_FACE);
		glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);
		glClearColor(0.0f, 0.0f, 0.0f, 0.0f);

		mainWindow = this;
		//Opening Sequence
		//*
		int frame = 0;
		Program openingShader = new Program("res/shaders/openingShaders/vert.gls", "res/shaders/openingShaders/frag.gls");
		GameObject openingObject = new GameObject("res/enignets.obj");
		Camera openingCamera = new Camera(70, 0.01f, 1000f);
		openingCamera.orderOfTransformations = new int[] {0, 2, 1};
		openingCamera.translate(0f, 0f, -5f);
		openingCamera.pitch((float)  Math.PI/2);

		while (openingCamera.getPitch() < (float) (3*Math.PI/2) && keys[GLFW_KEY_SPACE] != 1) {
			openingCamera.setPos(0f, 0f, (float) -(13-10f*Math.pow(150f-(float) frame, 6f)/1.1390625e+13));
			openingCamera.pitch(0.011f);
			update();
			frame += 1;
			openingShader.enable();
			openingShader.shaders[2].uniforms[0].set(new float[] {(float) frame});
			openingShader.shaders[2].uniforms[1].set(new float[] {(float) EnigWindow.width, (float) EnigWindow.height});
			openingObject.render(openingCamera);
			Program.disable();
			glfwSwapBuffers(window);
			glfwPollEvents();
			checkGLError();
		}
		frame = 286;
		openingCamera.setPos(0f, 0f, (float) -(13-10f*Math.pow(150f-(float) frame, 6f)/1.1390625e+13));
		openingCamera.setPitch((float) (3*Math.PI/2));
		update();
		openingShader.enable();
		openingShader.shaders[2].uniforms[0].set(new float[] {(float) frame});
		openingShader.shaders[2].uniforms[1].set(new float[] {(float) EnigWindow.width, (float) EnigWindow.height});
		openingObject.render(openingCamera);
		Program.disable();
		glfwSwapBuffers(window);
		glfwPollEvents();
		checkGLError();
		while (frame < 400 && keys[GLFW_KEY_SPACE] != 1) {
			++frame;
			update();
			openingShader.enable();
			openingShader.shaders[2].uniforms[0].set(new float[]{(float) frame});
			openingShader.shaders[2].uniforms[1].set(new float[]{(float) EnigWindow.width, (float) EnigWindow.height});
			openingObject.render(openingCamera);
			Program.disable();
			EnigWindow.mainWindow.resetOffsets();
			glfwSwapBuffers(EnigWindow.mainWindow.window);
			glfwPollEvents();
			EnigWindow.checkGLError();
		}
		//*/
	}
	
	public static void checkGLError() {
		int error;
		while ((error = glGetError()) != GL_NO_ERROR) {
			String errorInfo = "idk";
			switch (error) {
				case GL_NO_ERROR:
					errorInfo = "No error";
				case GL_INVALID_ENUM:
					errorInfo = "Invalid enum";
				case GL_INVALID_VALUE:
					errorInfo = "Invalid value";
				case GL_INVALID_OPERATION:
					errorInfo = "Invalid operation";
				case GL_STACK_OVERFLOW:
					errorInfo = "Stack overflow";
				case GL_STACK_UNDERFLOW:
					errorInfo = "Stack underflow";
				case GL_OUT_OF_MEMORY:
					errorInfo = "Out of memory";
				case GL_TABLE_TOO_LARGE:
					errorInfo = "Table too large";
				case GL_INVALID_FRAMEBUFFER_OPERATION:
					errorInfo = "Invalid framebuffer operation";
				default:
					errorInfo = "I rly dik";
			}
			System.out.println("id: " + error + " : " + errorInfo);
			throw new RuntimeException("id: " + error + " : " + errorInfo);
		}
	}
	
	public void update() {
		sync(60);
		for (int i = 0; i < keys.length; i++) {
			if (keys[i] == 1) {
				++keys[i];
			}
		}
		glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT); // clear the framebuffer
	}
	
	public void resetOffsets() {
		cursorXOffset = 0;
		cursorYOffset = 0;
		++framesSinceLastSecond;
		long time = System.nanoTime();
		if (time - lastSecond >= 1000000000) {
			lastSecond = time;
			lastFrameCount = framesSinceLastSecond;
			framesSinceLastSecond = 0;
			if (printFrames) {
				System.out.println("FPS: " + lastFrameCount);
			}
		}
	}
	
	public boolean toggleCursorInput() {
		if (inputEnabled) {
			glfwSetInputMode(window, GLFW_CURSOR, GLFW_CURSOR_NORMAL);
		}else {
			glfwSetInputMode(window, GLFW_CURSOR, GLFW_CURSOR_DISABLED);
		}
		inputEnabled = !inputEnabled;
		return inputEnabled;
	}
	
	private long variableYieldTime, lastTime;
	
	/**
	 * An accurate sync method that adapts automatically
	 * to the system it runs on to provide reliable results.
	 *
	 * @param fps The desired frame rate, in frames per second
	 * @author kappa (On the LWJGL Forums)
	 */
	private void sync(int fps) {
		if (fps <= 0) return;
		
		long sleepTime = 1000000000 / fps; // nanoseconds to sleep this frame
		// yieldTime + remainder micro & nano seconds if smaller than sleepTime
		long yieldTime = Math.min(sleepTime, variableYieldTime + sleepTime % (1000*1000));
		long overSleep = 0; // time the sync goes over by
		
		try {
			while (true) {
				long t = System.nanoTime() - lastTime;
				
				if (t < sleepTime - yieldTime) {
					Thread.sleep(1);
				}else if (t < sleepTime) {
					// burn the last few CPU cycles to ensure accuracy
					Thread.yield();
				}else {
					overSleep = t - sleepTime;
					break; // exit while loop
				}
			}
		} catch (InterruptedException e) {
			e.printStackTrace();
		}finally{
			lastTime = System.nanoTime() - Math.min(overSleep, sleepTime);
			
			// auto tune the time sync should yield
			if (overSleep > variableYieldTime) {
				// increase by 200 microseconds (1/5 a ms)
				variableYieldTime = Math.min(variableYieldTime + 200*1000, sleepTime);
			}
			else if (overSleep < variableYieldTime - 200*1000) {
				// decrease by 2 microseconds
				variableYieldTime = Math.max(variableYieldTime - 2*1000, 0);
			}
		}
	}
}

File: src/main/kotlin/engine/opengl/EnigWindow.kt
package engine.opengl

import engine.input.InputHandler
import engine.makeGLFWImage
import engine.opengl.shaders.ShaderProgram
import org.lwjgl.glfw.Callbacks
import org.lwjgl.system.MemoryUtil
import java.lang.RuntimeException
import org.joml.Vector2i
import org.lwjgl.glfw.GLFWImage
import org.lwjgl.glfw.GLFW.*
import org.lwjgl.opengl.GL11.*

private fun getPrimaryMonitorDimensions() : Vector2i {
	val vidMode = glfwGetVideoMode(glfwGetPrimaryMonitor())
	if (vidMode != null) {
		return Vector2i(vidMode.width(), vidMode.height())
	} else {
		throw RuntimeException("unable to find the dimensions of the primary monitor")
	}
}

@Suppress("MemberVisibilityCanBePrivate", "unused")
open class EnigWindow {
	open val id: Long
	var width = 0
		private set
	var height = 0
		private set
	open val aspectRatio
		get() = width.toFloat() / height.toFloat()
	open var fps = 144
	open var inputEnabled = false
		set(value) {
			glfwSetInputMode(id, GLFW_CURSOR, if (value) GLFW_CURSOR_DISABLED else GLFW_CURSOR_NORMAL)
			field = value
		}

	open val inputHandler : InputHandler

	/**
	 * creates a window
	 * @param windowSize size of window
	 * @param title name of the window
	 * @param decorated if the window should be decorated
	 * @param resizeable if the window should be resizeable
	 */
	constructor(windowSize : Vector2i, title: String, preset : GLContextPreset = GLContextPreset.standard,
	            decorated: Boolean = true, resizeable: Boolean = true) {

		// Configure GLFW
		glfwDefaultWindowHints()
		glfwWindowHint(GLFW_SAMPLES, 4)
		glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE)
		glfwWindowHint(GLFW_RESIZABLE, if (resizeable) 1 else 0)
		glfwWindowHint(GLFW_DECORATED, if (decorated) 1 else 0)
		glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 4)
		glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 6)
		glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE)
		glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE)

		// Create the window
		width = windowSize.x()
		height = windowSize.y()
		id = glfwCreateWindow(width, height, title, MemoryUtil.NULL, MemoryUtil.NULL) //fullscreen
		if (id == MemoryUtil.NULL) {
			throw RuntimeException("Failed to create the GLFW window")
		}

		inputHandler = InputHandler(id)

		glfwSetInputMode(id, GLFW_CURSOR, GLFW_CURSOR_NORMAL)

		// Make the window visible
		glfwShowWindow(id)
		if (resizeable) {
			glfwSetWindowSizeCallback(id) { _: Long, w: Int, h: Int ->
				width = w
				height = h
				setViewport()
			}
		}

		setContext()
		preset.createContext()
		// Enable v-sync
		//glfwSwapInterval(1)
	}

	/**
	 * creates a window
	 * @param w width
	 * @param h height
	 * @param title name of the window
	 * @param decorated if the window should be decorated
	 * @param resizeable if the window should be resizeable
	 */
	constructor(w: Int, h: Int, title: String, preset : GLContextPreset = GLContextPreset.standard,
	            decorated: Boolean = true, resizeable: Boolean = true)
			: this(Vector2i(w, h), title, preset, decorated, resizeable)

	/**
	 * creates a window
	 * @param w width
	 * @param h height
	 * @param title name of the window
	 * @param title name of the window
	 * @param decorated if it should be decorated or not
	 * @param resizeable if the window should be resizeable
	 */
	constructor(w : Int, h : Int, title : String, iconPath : String,
	            preset : GLContextPreset = GLContextPreset.standard, decorated: Boolean = false,
	            resizeable: Boolean = true)
			: this(w, h, title, preset, decorated, resizeable) {
		setIcon(iconPath)
	}

	/**
	 * creates an undecorated window
	 * @param title name of the window
	 * @param decorated if the window should be decorated or not
	 * @param resizeable if the window should be resizeable or not
	 */
	constructor(title : String, preset : GLContextPreset = GLContextPreset.standard, decorated : Boolean = false,
	            resizeable : Boolean = true)
			: this(getPrimaryMonitorDimensions(), title, preset, decorated, resizeable)

	/**
	 * creates an undecorated window
	 * @param title name of the window
	 * @param iconPath path to the icon for the window
	 * @param decorated if the window should be decorated or not
	 * @param resizeable if the window should be resizeable or not
	 */
	constructor(title : String, iconPath : String, preset : GLContextPreset = GLContextPreset.standard,
	            decorated : Boolean = false, resizeable : Boolean = true)
			: this(title, preset, decorated, resizeable) {
		setIcon(iconPath)
	}

	open fun setContext() = glfwMakeContextCurrent(id)

	open fun<T, R> runFrame(runData : T, drawingFun : (T) -> R) : R {
		setContext()

		val ret = drawingFun(runData)

		update()
		return ret
	}

	open fun runView(drawingFun : (frameBirth : Long, dtime : Float) -> Boolean) {
		setContext()
		val startTime = System.nanoTime()
		var birth = 0L
		var delta : Float
		do {
			val newTime = System.nanoTime() - startTime
			delta = (newTime - birth).toFloat() / 1e9f
			birth = newTime

			val shouldExit = drawingFun(birth, delta)
			update()
		} while (!glfwWindowShouldClose(id) && !shouldExit)
	}

	/**
	 * close the window, delete all resources, close openGL, openAL, and clean things up
	 */
	open fun terminate() {
		// Free the window callbacks and destroy the window
		Callbacks.glfwFreeCallbacks(id)
		glfwDestroyWindow(id)
	}

	/**
	 * syncs the fps and
	 */
	open fun update() {
		//sync(fps)
		inputHandler.update()
		ShaderProgram.disable()
		glfwSwapBuffers(id)
		glfwPollEvents()
		checkGLError()
	}

	private var variableYieldTime: Long = 0
	private var lastTime: Long = 0

	// TODO ensure that this isn't just wasting time
	/**
	 * An accurate sync method that adapts automatically
	 * to the system it runs on to provide reliable results.
	 *
	 * @param fps The desired frame rate, in frames per second
	 * @author kappa (On the LWJGL Forums)
	 */
	private fun sync(fps: Int) {
		if (fps <= 0) return
		val sleepTime = (1000000000 / fps).toLong() // nanoseconds to sleep this frame
		// yieldTime + remainder micro & nano seconds if smaller than sleepTime
		val yieldTime = sleepTime.coerceAtMost(variableYieldTime + sleepTime % (1000 * 1000))
		var overSleep: Long = 0 // time the sync goes over by
		try {
			while (true) {
				val t = System.nanoTime() - lastTime
				if (t < sleepTime - yieldTime) {
					//Thread.sleep(1);
				} else if (t < sleepTime) {
					// burn the last few CPU cycles to ensure accuracy
					Thread.yield()
				} else {
					overSleep = t - sleepTime
					break // exit while loop
				}
			}
		} finally {
			lastTime = System.nanoTime() - overSleep.coerceAtMost(sleepTime)

			// auto tune the time sync should yield
			if (overSleep > variableYieldTime) {
				// increase by 200 microseconds (1/5 a ms)
				variableYieldTime = sleepTime.coerceAtMost(variableYieldTime + 200 * 1000)
			} else if (overSleep < variableYieldTime - 200 * 1000) {
				// decrease by 2 microseconds
				variableYieldTime = (variableYieldTime - 2 * 1000).coerceAtLeast(0)
			}
		}
	}

	open fun setViewport() {
		val frameBufferWidth = IntArray(1)
		val frameBufferHeight = IntArray(1)
		glfwGetFramebufferSize(id, frameBufferWidth, frameBufferHeight)
		setViewport(0, 0, frameBufferWidth[0], frameBufferHeight[0])
	}

	open fun setViewport(x : Int, y : Int, width : Int, height : Int) {
		glViewport(x, y, width, height)
	}

	open fun setIcon(imagePath: String) {
		val image = makeGLFWImage(imagePath)
		val buffer = GLFWImage.malloc(1)
		buffer.put(0, image)
		glfwSetWindowIcon(id, buffer)
	}
}